nanoAOD/plugins/HadTruthProducer.cc

```diff
@@ -73,12 +79,49 @@ HadTruthProducer::produce( edm::Event& iEvent, const edm::EventSetup& iSetup)
     if (nmatched == 2) {
       isHadFrom(trueHad, 6, count, hadFromQuark, hadFromTop);
       matchHad.insert({trueHad.get(), &cand});
-    }
+      matchedGen.push_back(trueHad.get());
+    } else { matchedGen.push_back(nullptr); }
 
     isHadFromTsb.push_back(hadFromQuark);
     isHadFromTop.push_back(hadFromTop);
   }
   
+  for (auto indices : *hadronIndices) {
+    if (abs((*hadronCands)[indices[0]].pdgId()) != 5122) continue;
+    
+    int nmatched = 0;
+    int count=0;
+    int hadFromQuark=-99;
+    bool hadFromTop=false;
+    
+    std::vector<reco::GenParticleRef> momlist;    
+    
+    for (auto i = 1; i <= 2; i++){
+      auto dau = (*hadronCands)[indices[i]];
+      auto gen_dau = matchedGen[indices[i]];
+      if (gen_dau == nullptr) continue;
+
+      auto mother = gen_dau->motherRefVector();
+      auto im = mother.begin();
+      while (im < mother.end()){
+        if (abs((*im)->pdgId()) == 5122){
+          momlist.push_back(*im);
+          nmatched ++;
+          break;
+        }
+        im++;
+      }
+    }
+
+    if (nmatched == 2) {
+      if (momlist[0] != momlist[1]) { nmatched = 1; }
```

Comment:
loop through mom list and check they're all the same, or else do something like trueHad for tracking above, where we fill a pointer with the first hadron mother, then check subsequent daughters have the same mother, break out of the loop false if not (that would also mean we don't need nmatched checking here)